src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIRequestSmelter.java

```diff
@@ -218,6 +223,140 @@ private IAIState accelerateFurnaces()
+        final int amountOfFuelInInv = InventoryUtils.getItemCountInItemHandler(worker.getInventoryCitizen(), item -> FurnaceTileEntity.isFuel(item) && possibleFuels.stream().anyMatch(candidate -> ItemStackUtils.compareItemStacksIgnoreStackSize(candidate, item)));
+
+        if (amountOfFuelInBuilding + amountOfFuelInInv <= 0 && !getOwnBuilding().hasWorkerOpenRequestsOfType(worker.getCitizenData(), TypeToken.of(StackList.class)) && currentRecipeStorage != null && currentRecipeStorage.getIntermediate() == Blocks.FURNACE)
+        {
+            worker.getCitizenData().createRequestAsync(new StackList(possibleFuels, COM_MINECOLONIES_REQUESTS_BURNABLE, STACKSIZE, 1));
+            return getState();
+        }
+
+        for (final BlockPos pos : getOwnBuilding().getFurnaces())
+        {
+            if (WorldUtil.isBlockLoaded(world, pos))
+            {
+                final TileEntity entity = world.getTileEntity(pos);
+                if (entity instanceof FurnaceTileEntity)
+                {
+                    final FurnaceTileEntity furnace = (FurnaceTileEntity) entity;
+                    if (!furnace.isBurning() && (hasSmeltableInFurnaceAndNoFuel(furnace) || hasNeitherFuelNorSmeltAble(furnace)) && currentRecipeStorage != null && currentRecipeStorage.getIntermediate() == Blocks.FURNACE) 
+                    {
+                        if (amountOfFuelInBuilding > 0 && amountOfFuelInInv == 0)
+                        {
+                            needsCurrently = new Tuple<>(item -> FurnaceTileEntity.isFuel(item) && possibleFuels.stream().anyMatch(candidate -> ItemStackUtils.compareItemStacksIgnoreStackSize(candidate, item)), STACKSIZE);
+                            return GATHERING_REQUIRED_MATERIALS;
+                        }
+
+                        fuelPos = pos;
+                        if(preFuelState == null)
+                        {
+                            preFuelState = getState();
+                        }
+                        return ADD_FUEL_TO_FURNACE;
+                    }
+                }
+            }
+        }
+        return getState();
+    }
+
+    /**
+     * Add furnace fuel when necessary
+     * @return
+     */
+    private IAIState addFuelToFurnace()
+    {
+        final List<ItemStack> possibleFuels = getActivePossibleFuels();
+        final int amountOfFuelInBuilding = InventoryUtils.getItemCountInProvider(getOwnBuilding(), item -> FurnaceTileEntity.isFuel(item) && possibleFuels.stream().anyMatch(candidate -> ItemStackUtils.compareItemStacksIgnoreStackSize(candidate, item)));
+        final int amountOfFuelInInv = InventoryUtils.getItemCountInItemHandler(worker.getInventoryCitizen(), item -> FurnaceTileEntity.isFuel(item) && possibleFuels.stream().anyMatch(candidate -> ItemStackUtils.compareItemStacksIgnoreStackSize(candidate, item)));
+
+        if (amountOfFuelInBuilding > 0 && amountOfFuelInInv == 0)
```

Comment:
if both are 0, I could imagine this here being a problem

Reply:
Also I think we can optimize this (only check one if the other is x, etc).
Especially since we walk to the block, we definitely want to do the expensive count operations only after arriving. before we at most care "are there any"

Reply:
The problem is that we don't want to want to walk to the block unless the counts are where we need them to be. At this point we should be sitting at the building, and only if we have the fuel to put in the furnace should we walk. If we do the gather materials, we'd be walking to the rack with the fuel. 

I can do the move of the building check to only happen if we have none in inventory, and make sure we don't get to the walkTo if we have no fuel in inventory.

Reply:
we can though just check for "hasInInv" instead of counting. And only count inBuilding if the fuelInInv is 0

Reply:
By "hasInInv" you mean InventoryUtils.hasItemIn<Provider/Handler>?

Reply:
yes

Reply:
Looking at all this again fresh, I'm not sure we actually need the count here.